compiler/luci-interpreter/src/kernels/BroadcastTo.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kernels/BroadcastTo.h"
+#include "kernels/Utils.h"
+
+#include "PALBroadcastTo.h"
+
+#include <stdexcept>
+#include <iostream>
+
+namespace luci_interpreter
+{
+
+namespace kernels
+{
+
+// TODO Extract this function to Utils.h
+Shape extractShapeFromTensor(const Tensor *tensor)
+{
+  Shape shape(tensor->shape().num_elements());
+  if (tensor->element_type() == DataType::S32)
+  {
+    const auto *shape_data = tensor->data<int32_t>();
+    for (int i = 0; i < tensor->shape().num_elements(); ++i)
+    {
+      shape.dim(i) = shape_data[i];
+    }
+  }
+  else if (tensor->element_type() == DataType::S64)
+  {
+    const auto *shape_data = tensor->data<int64_t>();
+    for (int i = 0; i < tensor->shape().num_elements(); ++i)
+    {
+      shape.dim(i) = static_cast<int32_t>(shape_data[i]);
+      // Check value overflow
+      LUCI_INTERPRETER_CHECK(static_cast<int64_t>(shape.dim(i)) == shape_data[i]);
+    }
+  }
+  else
+  {
+    LUCI_INTERPRETER_CHECK(false);
+  }
+  return shape;
+}
+
+BroadcastTo::BroadcastTo(const Tensor *input, const Tensor *shape, Tensor *output)
+  : Kernel({input, shape}, {output})
+{
+}
+
+void BroadcastTo::configure()
+{
+  Shape output_shape = extractShapeFromTensor(shape());
+
+  // Check if output shape is broadcastable from input shape
+  int input_rank = input()->shape().num_dims();
+  int output_rank = output_shape.num_dims();
+  int extending_rank = output_rank - input_rank;
+
+  for (int idx = 0; idx < input_rank; ++idx)
+  {
+    LUCI_INTERPRETER_CHECK(input()->shape().dim(idx) == 1 ||
+                           input()->shape().dim(idx) == output_shape.dim(extending_rank + idx));
+  }
```

Comment:
@seanshpark 

> Have to checked tflite kernel source codes with BroadcastTo shape input?

I added code to check broadcastable value from like [broadcast_to.cc of tflite-kernels](https://github.com/tensorflow/tensorflow/blob/1b8f5c396f0c016ebe81fe1af029e6f205c926a4/tensorflow/lite/kernels/broadcast_to.cc#L60-L76).

new code checks the shape input value here.

>  is there no cases where shape_data[i] < 0 ?

If the shape input is negative dim values, it is checked that it is not broadcastable because it is different from the dim values of the input shape (If the dim value of input shape is positive).


> Then, what happens if the dim value of the input shape is negative?

In the `ONE` recipe, if the dim value of the input shape is set to a negative number, the following protobuf-related ERROR occurs, so it cannot be possible in the recipe.

```
[ 13%] Generating BroadcastTo_000.tflite
[libprotobuf ERROR /home/leeseunghui/workspace/ONE/externals/PROTOBUF/src/google/protobuf/text_format.cc:288] Error parsing text-format tflchef.ModelRecipe: 4:16: Expected integer, got: "-2"
ERROR: Failed to parse recipe 'BroadcastTo_000.recipe'
make[2]: *** [compiler/luci/tests/CMakeFiles/luci_testfiles.dir/build.make:3047: compiler/luci/tests/BroadcastTo_000.tflite] Error 255
make[1]: *** [CMakeFiles/Makefile2:6671: compiler/luci/tests/CMakeFiles/luci_testfiles.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```


And, I additionally added the test code to check the broadcastable.

Reply:
1/ please share the recipe file. 
2/ why have you used `-2` ?

Reply:
> 1/ please share the recipe file.
```
# BroadcastTo Op supports op version 2 and 3.
# BroadcastTo of Float type is version 2 and Quantized BroadcastTo is version 3.
# We are using tflite kernels for float values. so, the version should be 2.
# refer https://github.com/tensorflow/tensorflow/pull/46224/files
operand {
  name: "bc_input"
  type: FLOAT32
  shape { dim: "-2" dim: "3" }
}
operand {
  name: "bc_shape"
  type: INT32
  shape { dim: 3 }
  filler { tag: "explicit" arg: "1" arg: "-2" arg: "3" }
}
operand {
  name: "bc_ofm"
  type: FLOAT32
  shape { dim: "1" dim: "-2" dim: "3" }
}
operation {
  type: "BroadcastTo"
  input: "bc_input"
  input: "bc_shape"
  output: "bc_ofm"
  version: 2
}
input: "bc_input"
output: "bc_ofm"
```

> 2/ why have you used -2 ? what is the meaning of -2 ?

I just used `-2` to test if it is possible when dim values of `input` and `shape` are both negative. (In this case, it is set to any number that is judged to be broadcastable in the broadcast_to.cc of tflite kernel.)

Reply:
minus values should work. have you checked other recipe files?

> I just used -2 to test if

As I understand you expect `-2` should also work. is this correct?